demos/instr.h

```diff
@@ -77,4 +77,58 @@ inline void JumpToAfterSpeculation() {
   asm volatile("b afterspeculation");
 }
 #endif
+
+#ifdef __i386__
+// Returns the original value of FS and sets the new value.
+__attribute__((always_inline))
+inline static int ExchangeFS(int input) {
+  asm volatile(
+      "mov %%fs, %%eax\n"
+      "movl %1, %%fs\n"
+      "movl %%eax, %0\n"
+      :"=r"(input):"r"(input):"eax");
```

Comment:
(here and below)

I think you can just list `input` once -- ironically, as an output -- by prefixing it with `+` to say it's read and written.

https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#OutputOperands

let's also use a temp variable that we declare and assign as an output rather than statically clobber `eax`.

Reply:
There is probably a bug in the compiler. When I write it this way:

__attribute__((always_inline))
inline static int ExchangeFS(int input) {
  int tmp;
  asm volatile(
      "mov %%fs, %1\n"
      "movl %0, %%fs\n"
      "movl %1, %0\n"
      :"=r"(input):"r"(tmp));
  return input;
}

__attribute__((always_inline))
inline static int ExchangeES(int input) {
  int tmp;
  asm volatile(
      "mov %%es, %1\n"
      "movl %0, %%es\n"
      "movl %1, %0\n"
      :"=r"(input):"r"(tmp));
  return input;
}

The inlined assembly ends up like this:

afterspeculation:
# 0 "" 2
#NO_APP
        xorl    %eax, %eax
#APP
# 90 "instr.h" 1
        mov %fs, %eax
movl %eax, %fs
movl %eax, %eax
# 0 "" 2
#NO_APP
        xorl    %eax, %eax
#APP
# 102 "instr.h" 1
        mov %es, %eax
movl %eax, %es
movl %eax, %eax
# 0 "" 2
#NO_APP
        movl    -1076(%ebp), %ebx

Which means that the backup value from the first call is just discarded by the compiler (g++ 8.3.0). It works only when compiled without optimizations.

Reply:
I'll try to implement it using pointers. That should work more reliably.